library/nist_kw.c

```diff
@@ -0,0 +1,654 @@
+                            mbedtls_cipher_id_t cipher,
+                            const unsigned char *key,
+                            unsigned int keybits,
+                            const int is_wrap )
+{
+    int ret;
+    const mbedtls_cipher_info_t *cipher_info;
+
+
+    cipher_info = mbedtls_cipher_info_from_values( cipher,
+                                                   keybits,
+                                                   MBEDTLS_MODE_ECB );
+    if( cipher_info == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    if( cipher_info->block_size != 16 )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    /*
+      * SP 800-38F currently defines AES cipher as the only block cipher allowed:
+      * "For KW and KWP, the underlying block cipher shall be approved, and the block size shall be
+      *  128 bits. Currently, the AES block cipher, with key lengths of 128, 192, or 256 bits, is the only
+      *  block cipher that fits this profile."
+      *  Currently we don't support other 128 bit block ciphers for key wrapping, such as Camellia and Aria.
+      */
+     if( cipher != MBEDTLS_CIPHER_ID_AES )
+         return( MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE );
+
+     mbedtls_cipher_free( &ctx->cipher_ctx );
+
+    if( ( ret = mbedtls_cipher_setup( &ctx->cipher_ctx, cipher_info ) ) != 0 )
+        return( ret );
+
+    if( ( ret = mbedtls_cipher_setkey( &ctx->cipher_ctx, key, keybits,
+                                       is_wrap ? MBEDTLS_ENCRYPT : MBEDTLS_DECRYPT )
+                                     ) != 0 )
+    {
+        return( ret );
+    }
+
+    return( 0 );
+}
+
+/*
+ * Free context
+ */
+void mbedtls_nist_kw_free( mbedtls_nist_kw_context *ctx )
+{
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_nist_kw_context ) );
+}
+
+/*
+ * KW-AE as defined in SP 800-38F section 6.2
+ * KWP-AE as defined in SP 800-38F section 6.3
+ */
+int mbedtls_nist_kw_wrap( mbedtls_nist_kw_context *ctx, mbedtls_nist_kw_mode_t mode,
+                          const unsigned char *input, size_t in_len,
+                          unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    size_t semiblocks = 0;
+    size_t s ;
+    size_t olen, i, padlen = 0;
+    uint64_t t = 0;
+    unsigned char outbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char inbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char *R2 = output + KW_SEMIBLOCK_LENGTH;
+    unsigned char *A = output ;
+    const uint32_t ICV1[] = NIST_KW_ICV1;
+    const uint32_t ICV2[] = NIST_KW_ICV2;
+
+    /*
+     * Generate the String to work on
+     */
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+        /*
+         * According to SP800-38F Table 1, the plaintext length for KW
+         * must be between 2 to 2^54-1 semiblocks inclusive.
+         */
+        if( ( in_len < 8 )
+#if SIZE_MAX > 0xFFFFFFFF
+                || ( in_len > 0x1FFFFFFFFFFFFF8 )
```

Comment:
This indentation of `||` looks strange to me: it isn't aligned with anything, so it doesn't tell me about where this line fits in the overall expression. I'd prefer to align it either with the first non-blank character at the same level of nesting, i.e. at the start of `( in_len < 8 )`.

On a related note, I'd prefer not to have parentheses when they're obviously not needed. Parentheses clutter up an expression, they tell the reader that there's something to be careful about. They're useful when the parsing of the expression is not obvious, even if they aren't needed for correctness. But some things _are_ obvious and shouldn't draw the attention of the reader, such as
```
    comparison || comparison
    multiplication + multiplication
```